File: scalang/src/main/scala/com/kapilkoju/autopos/kernel/api/GetApi.scala
package com.kapilkoju.autopos.kernel.api

import com.codahale.metrics.annotation.Timed
import com.kapilkoju.autopos.kernel.domain.AuditableEntity
import org.springframework.http.HttpStatus.{NOT_FOUND, OK}
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.{GetMapping, PathVariable}

trait GetApi[T <: AuditableEntity] {
  this: Api[T] =>

  @GetMapping(value = Array("/{id}"))
  @Timed
  def get(@PathVariable("id") entityOp: T): ResponseEntity[_ <: T] = {
    Option(entityOp) match {
      case Some(entity) => new ResponseEntity(entity, OK)
      case None => new ResponseEntity(NOT_FOUND)
    }
  }
}

File: src/main/java/com/kapilkoju/autopos/kernel/api/GetApi.kt
package com.kapilkoju.autopos.kernel.api

import com.codahale.metrics.annotation.Timed
import com.kapilkoju.autopos.kernel.domain.AuditableEntity
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable

interface GetApi<T : AuditableEntity> : Api<T> {

    @GetMapping(value = ["/{id}"])
    @Timed
    fun get(@PathVariable("id") entityOp: T?): ResponseEntity<T> {

        return if (entityOp != null) {
            ResponseEntity(entityOp, HttpStatus.OK)
        } else {
            ResponseEntity(HttpStatus.NOT_FOUND)
        }
    }
}
